src/tablenavigation.js

```diff
@@ -0,0 +1,515 @@
+		}
+
+		return true;
+	}
+
+	/**
+	 * Moves the selection from the given table cell in the specified direction.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} tableCell The table cell to start the selection navigation.
+	 * @param {'left'|'up'|'right'|'down'} direction Direction in which selection should move.
+	 */
+	_navigateFromCellInDirection( tableCell, direction ) {
+		const model = this.editor.model;
+
+		const table = findAncestor( 'table', tableCell );
+		const tableMap = [ ...new TableWalker( table, { includeSpanned: true } ) ];
+		const { row: lastRow, column: lastColumn } = tableMap[ tableMap.length - 1 ];
+
+		const currentCellInfo = tableMap.find( ( { cell } ) => cell == tableCell );
+		let { row, column } = currentCellInfo;
+
+		switch ( direction ) {
+			case 'left':
+				column--;
+				break;
+
+			case 'up':
+				row--;
+				break;
+
+			case 'right':
+				column += currentCellInfo.colspan;
+				break;
+
+			case 'down':
+				row += currentCellInfo.rowspan;
+				break;
+		}
+
+		const isOutsideVertically = row < 0 || row > lastRow;
+		const isBeforeFirstCell = column < 0 && row <= 0;
+		const isAfterLastCell = column > lastColumn && row >= lastRow;
+
+		// Note that if the last table cell is row-spanned then isAfterLastCell will never be true but we don't know
+		// if user was navigating on the last row or not, so let's allow him to stay in the table.
+
+		if ( isOutsideVertically || isBeforeFirstCell || isAfterLastCell ) {
+			model.change( writer => {
+				writer.setSelection( writer.createRangeOn( table ) );
+			} );
+
+			return;
+		}
+
+		if ( column < 0 ) {
+			column = lastColumn;
+			row--;
+		} else if ( column > lastColumn ) {
+			column = 0;
+			row++;
+		}
+
+		const cellToSelect = tableMap.find( cellInfo => cellInfo.row == row && cellInfo.column == column ).cell;
+		const isForward = [ 'right', 'down' ].includes( direction );
+		const positionToSelect = model.createPositionAt( cellToSelect, isForward ? 0 : 'end' );
+
+		model.change( writer => {
+			writer.setSelection( positionToSelect );
+		} );
+	}
+}
+
+// Returns 'true' if provided key code represents one of the arrow keys.
+//
+// @private
+// @param {Number} keyCode
+// @returns {Boolean}
+function isArrowKeyCode( keyCode ) {
+	return keyCode == keyCodes.arrowright ||
+		keyCode == keyCodes.arrowleft ||
+		keyCode == keyCodes.arrowup ||
+		keyCode == keyCodes.arrowdown;
+}
+
+// Returns direction name from `keyCode`.
+//
+// @private
+// @param {Number} keyCode
+// @param {Boolean} isLtrContent The content language direction.
+// @returns {'left'|'up'|'right'|'down'} Arrow direction.
+function getDirectionFromKeyCode( keyCode, isLtrContent ) {
+	switch ( keyCode ) {
+		case keyCodes.arrowleft: return isLtrContent ? 'left' : 'right';
+		case keyCodes.arrowright: return isLtrContent ? 'right' : 'left';
+		case keyCodes.arrowup: return 'up';
+		case keyCodes.arrowdown: return 'down';
```

Comment:
Please format it so each `return` is in new line (code style consistency with existing code).